<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>rxjs</title>
</head>
<body>
    <h1>在集合console.log来查看代码的演示效果</h1>
</body>
<script src="https://cdn.bootcdn.net/ajax/libs/rxjs/6.2.0/rxjs.umd.min.js"></script>
<script>
    //  cdn引用的直接调用
    // rxjs.interval(500).pipe(rxjs.operators.take(4)).subscribe(console.log)

    //  catchError(project : function): Observable
    //  捕获错误
    
    //  单参数
    // const {interval} = rxjs
    // const {take,map,catchError} = rxjs.operators

    // interval(1000).pipe(
    //     take(6),
    //     map(x =>{
    //         if(x === 4){
    //             throw new Error('unlucky number 4')
    //         }else{
    //             return x
    //         }
    //     }),
    //     //  捕获异常后抛出8
    //     catchError(error => of(8))
    // ).subscribe(x => console.log(x))


    // 双参数
    const {interval} = rxjs
    const {take,map,catchError} = rxjs.operators
    interval(1000).pipe(
        take(6),
        map(x=>{
            if(x === 4){
                throw new Error('unlucky number 4')
            }else{
                return x
            }
        }),
        //  双参数，返回错误发生前捕获的内容,再次重试
        catchError((err,caught$) => caught$),
        take(10)
    ).subscribe(x=>console.log(x))
</script>
</html>